src/main/java/com/work/management/service/employee/impl/EmployeeServiceImpl.java

```diff
@@ -51,4 +56,39 @@ public EmployeeDto getEmployeeByUserName(String username) {
     BeanUtils.copyProperties(employee.get(), employeeDto);
     return employeeDto;
   }
+
+  @Override
+  public EmployeeDto updateEmployeeEntity(EmployeeDto employeeDto) {
+    Optional<Employee> employee = employeeRepository.findById(employeeDto.getId());
+    if (!employee.isPresent()) {
+      ExceptionUtils
+          .throwEntityNotFoundException(
+              String
+                  .format("Employee with Id %d doesn't exists.", employeeDto.getId()));
+    }
+
+    BeanUtils.copyProperties(employeeDto, employee.get(), getNullFieldNames(employeeDto));
+    employeeRepository.save(employee.get());
```

Comment:
I don't think you need getNullFieldNames here... just write this function as...

```
 Optional<Employee> employee = employeeRepository.findById(employeeDto.getId());
    if (!employee.isPresent()) {
      ExceptionUtils
          .throwEntityNotFoundException(
              String
                  .format("Employee with Id %d doesn't exists.", employeeDto.getId()));
    }
Employee oldEmployee = employee.get();
if(employeeDto.getUsername!=null&& !oldEmployee.getUsername().equals(employeeDto.getUsername)|| employeeDto.getId()!=null&&!oldEmployee.getId().equals(employeeDto.getId())
{
     throw new BadRequestException("Can't update username or id.").
}
Employee newEmployee = new Employee();


BeanUtils.copyProperties(employeeDto, newEmployee);
newEmployee.setId(oldEmployee.getId());
newEmployee.setUserName(oldEmployee.getUsername);

employeeRepository.save(newEmployee);
BeanUtils.copyProperties(newEmployee, employeeDto);
    return employeeDto;
  }
```

Reply:
One more thing here...employeeDto object should be immutable...remove setters from it...Always create a new employeeDto object to and copy details to that new object and return that....Dto object should be immutable...they should be made through builders or constructor only.

Never update the instance which is coming from function parameter...remember this.. and update all your functions according to it...make this rule in your mind.

Reply:
Functions which don't update their function parameters are called pure functions. And these functions are easy to test, and less prone to errors..somebody might be using that function parameter parallely in some other thread...after calling your function..and you are changing that parameter..means you can screw his logic...

Reply:
Yup okay , much appreciated